honeybadgermpc/protocols/avid.py

```
@@ -0,0 +1,216 @@
+            stripes_list = [None] * self.input_size
+            mt_list = [None] * self.input_size
+            roothash_list = [None] * self.input_size
+            for i, m in enumerate(input_list):
+                stripes_list[i] = encode(k, self.n, m)
+                mt_list[i] = merkle_tree(stripes_list[i])
+                roothash_list[i] = mt_list[i][1]
+
+            # transpose for sending
+            stripes_list_per_party = [list(i) for i in zip(*stripes_list)]
+            for i in range(self.n):
+                branch_list = [None] * self.input_size
+                for j in range(self.input_size):
+                    branch_list[j] = get_merkle_branch(i, mt_list[j])
+                # send each person the column of stripes
+                self.send(i, (sid, 'VAL', roothash_list,
+                              branch_list, stripes_list_per_party[i]))
+
+        # counters and variables for msg
+        echo_set = set()
+        ready_set = set()
+        ready_sent = False
+        from_leader = None
+        # internal storage
+        my_stripes = [None] * self.input_size
+        my_roothash_list = [None] * self.input_size
+
+        while True:  # main recv loop
+            sender, msg = await self.recv()
+            if msg[1] == 'VAL' and from_leader is None:
+                # Validation
+                (_, _, roothash_list, branch_list, stripes_for_each) = msg
+                if sender != self.leader:
+                    logger.warning(
+                        f"[{pid}] VAL message from other than leader: {sender}")
+                    continue
+
+                # merkle tree verification
+                validation_fail_flag = False
+                for i in range(len(stripes_for_each)):
+                    # verify each entry in the stripes
+                    try:
+                        assert merkle_verify(self.n, stripes_for_each[i],
```

This is weird, if `merkle_verify` returns a bool then why don't you do an `if` instead of calling `assert`? Checking using asserts is bad when you want to really throw an exception because we disable the asserts when running the code using the `-O` flag.

---

This whole, `try` `except` can be replaced with `if` and `raise`

---

This is validation of input from untrusted parties. It should not be an `assert`.

---

So do we want to raise an error and stop this process or we don't add it to the set and keep getting new values?

---

Print the error, keep accepting new values (from the other, non-faulty, nodes).
This is reminding me we should have a unified "mark party as byzantine" system (aka, failure detector) since basically if a party is faulty in one place, we should probably ignore them in other contexts too

---

Resolved.